### PR TITLE
CORE-5260 new Merkle tree implementation

### PR DESCRIPTION
**Pull request by vlajos**
New Merkle tree implementation.
API: https://github.com/corda/corda-api/pull/435

### CODE REVIEW

File: libs/crypto/merkle/src/main/java/net/corda/crypto/merkle/package-info.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+@Export
+package net.corda.crypto.merkle;
+
+import org.osgi.annotation.bundle.Export;
```
Comment (blsemo) [inline libs/crypto/merkle/src/main/java/net/corda/crypto/merkle/package-info.java]: Should this file be here? This is an impl package, so we don't want to export it as a bundle that can be imported? Or does it need to be exported as it is not an OSGi component itself?

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                        if (item.first xor next.first == 1) {       // ... and they are a pair with the current
+                            newItems += Pair(                       // in the original tree, we create their parent.
+                                item.first / 2,
+                                digestProvider.nodeHash(treeDepth, item.second, next.second)
+                            )
+                            index += 2
+                            continue
+                        }
+                    }
+                    if (hashIndex >= hashes.size) {                 // We'll need and more hash to continue. So if
```
Comment (blsemo) [inline libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt]: Typo in the comment. The `and` doesn't make sense.

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                        return false                                // we do not have more, the proof is incorrect.
+                    }
+                                                                    // We pair the current element with a
+                                                                    // hash from the proof
+                    newItems += if (item.first and 1 == 0) {        // Even index means, that the item is on the left
```
Comment (blsemo) [inline libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt]:
```suggestion
                    newItems += if ((item.first and 1) == 0) {        // Even index means, that the item is on the left
```
Would suggest parenthesis to make reading easier - `1 == 0` threw me for a bit.

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                        Pair(
+                            item.first / 2,
+                            digestProvider.nodeHash(treeDepth, item.second, hashes[hashIndex++])
+                        )
+                    } else {                                        // Odd index means, that the item is on the right
+                        Pair(
+                            item.first / 2,
+                            digestProvider.nodeHash(treeDepth, hashes[hashIndex++], item.second)
```
Comment (blsemo) [inline libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt:76]: Is that correct? Should the leaf on the right not have an index that is one lower than the current leaf?

**Review by blsemo — COMMENTED** [review]
Looks generally good, a few qestions/remarks.

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt (hunk 4 above)
Comment (vlajos) [reply libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt:76]: This is roughly the case of H1 here: https://github.com/corda/corda-runtime-os/pull/1492/files#diff-7cfa00b3e9bdbc460bc6921838315aa67b4bac7544057a987d5b74347e2b3e69R55
item.first =1 and on the next level the new nodeHash will be on the 0th index.
Does this make sense?

File: libs/crypto/merkle/src/main/java/net/corda/crypto/merkle/package-info.java (hunk 1 above)
Comment (vlajos) [reply libs/crypto/merkle/src/main/java/net/corda/crypto/merkle/package-info.java]: CORE-5145 will make it accessible. I have removed this here.

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt (hunk 2 above)
Comment (vlajos) [reply libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt]: Thanks. Fixed.

File: libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt (hunk 3 above)
Comment (vlajos) [reply libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt]: Thanks. Fixed.

**Review by blsemo — DISMISSED** [review]
LGTM

**Review by blsemo — APPROVED** [review]
Re-approve after rebase

**Pull request merged**